File: clients/android/app/src/main/kotlin/com/lextures/android/core/auth/SessionsApi.kt
package com.lextures.android.core.auth

import com.lextures.android.core.network.ApiClient
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class ActiveSession(
    val id: String,
    @SerialName("createdAt") val createdAt: String,
    @SerialName("lastUsedAt") val lastUsedAt: String,
    @SerialName("deviceLabel") val deviceLabel: String,
    val location: String,
    @SerialName("authMethod") val authMethod: String,
    @SerialName("isCurrent") val isCurrent: Boolean,
)

@Serializable
private data class SessionsResponse(
    val sessions: List<ActiveSession>,
)

/** Device session list/revoke endpoints (`GET/DELETE /api/v1/me/sessions`). */
object SessionsApi {
    private val client = ApiClient()
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun fetchSessions(accessToken: String): List<ActiveSession> {
        val (body, _) = client.request(
            path = "/api/v1/me/sessions",
            accessToken = accessToken,
        )
        return json.decodeFromString(SessionsResponse.serializer(), body).sessions
    }

    suspend fun revokeSession(id: String, accessToken: String) {
        client.request(
            path = "/api/v1/me/sessions/$id",
            method = "DELETE",
            accessToken = accessToken,
        )
    }

    suspend fun revokeOtherSessions(accessToken: String) {
        client.request(
            path = "/api/v1/me/sessions",
            method = "DELETE",
            accessToken = accessToken,
        )
    }
}

File: clients/ios/Lextures/Core/Auth/SessionsAPI.swift
import Foundation

/// Device session list/revoke endpoints (`GET/DELETE /api/v1/me/sessions`).
enum SessionsAPI {
    static let client = APIClient()

    struct ActiveSession: Decodable, Identifiable, Equatable {
        var id: String
        var createdAt: String
        var lastUsedAt: String
        var deviceLabel: String
        var location: String
        var authMethod: String
        var isCurrent: Bool
    }

    private struct SessionsResponse: Decodable {
        var sessions: [ActiveSession]
    }

    static func fetchSessions(accessToken: String) async throws -> [ActiveSession] {
        let (data, _) = try await client.request(
            path: "/api/v1/me/sessions",
            authorized: true,
            accessToken: accessToken
        )
        return try LMSAPI.decode(SessionsResponse.self, from: data).sessions
    }

    static func revokeSession(id: String, accessToken: String) async throws {
        _ = try await client.request(
            path: "/api/v1/me/sessions/\(encodePath(id))",
            method: "DELETE",
            authorized: true,
            accessToken: accessToken
        )
    }

    static func revokeOtherSessions(accessToken: String) async throws {
        _ = try await client.request(
            path: "/api/v1/me/sessions",
            method: "DELETE",
            authorized: true,
            accessToken: accessToken
        )
    }

    private static func encodePath(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? value
    }
}
